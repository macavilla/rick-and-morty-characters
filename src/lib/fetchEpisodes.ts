import { EpisodeType } from "@/types/episode";

const fetchEpisodes = async (episodes: EpisodeType["id"][]) => {
  const response = await fetch(
    `https://rickandmortyapi.com/api/episode/${episodes}`
  );
  const data = await response.json();
  return Array.isArray(data) ? data : [data];
};

export default fetchEpisodes;
