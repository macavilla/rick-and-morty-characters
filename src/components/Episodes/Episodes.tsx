import { CharacterType } from "@/types/character";
import EpisodesList from "./EpisodesList/EpisodesList";
import { useEffect, useState } from "react";
import { get } from "https";
import { EpisodeType } from "@/types/episode";
import fetchEpisodes from "@/lib/fetchEpisodes";

interface EpisodesProps {
  character1Episode: CharacterType["episode"] | null;
  character2Episode: CharacterType["episode"] | null;
}

const getEpisodesIds = (episodes: CharacterType["episode"]) => {
  return episodes.map((ep) => Number(ep.split("/").pop()));
};
const Episodes = ({ character1Episode, character2Episode }: EpisodesProps) => {
  const [episodes1, setEpisodes1] = useState<EpisodeType[] | null>(null);
  const [episodes2, setEpisodes2] = useState<EpisodeType[] | null>(null);
  const [sharedEpisodes, setSharedEpisodes] = useState<EpisodeType[] | null>(
    null
  );

  useEffect(() => {
    if (!character1Episode && !character2Episode) return;

    if (character1Episode) {
      fetchEpisodes(getEpisodesIds(character1Episode)).then((episodes) => {
        setEpisodes1(episodes);
      });
    }

    if (character2Episode) {
      fetchEpisodes(getEpisodesIds(character2Episode)).then((episodes) => {
        setEpisodes2(episodes);
      });
    }

    if (character1Episode && character2Episode) {
      const char1EpisodeIds =
        character1Episode && getEpisodesIds(character1Episode);
      const char2EpisodeIds =
        character2Episode && getEpisodesIds(character2Episode);

      const sharedEpisodesIds = char1EpisodeIds?.filter((episodeId) =>
        char2EpisodeIds?.includes(episodeId)
      );

      fetchEpisodes(sharedEpisodesIds).then((episodes) => {
        setSharedEpisodes(episodes);
      });
    }
  }, [character1Episode, character2Episode]);

  return (
    <div className="md:flex justify-between">
      <div className="w-full md:w-1/3">
        <EpisodesList title="Character 1 episodes" episodes={episodes1} />
      </div>
      <div className="w-full md:w-1/3">
        <EpisodesList title="Shared episodes" episodes={sharedEpisodes} />
      </div>
      <div className="w-full md:w-1/3">
        <EpisodesList title="Character 2 episodes" episodes={episodes2} />
      </div>
    </div>
  );
};

export default Episodes;
