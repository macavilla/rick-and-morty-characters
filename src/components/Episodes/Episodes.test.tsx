import { screen, render, waitFor } from "@testing-library/react";
import {
  mockEpisodesProps,
  mockFetchEpisodes,
} from "./__mockData__/Episodes.mock";
import fetchEpisodes from "../../utils/fetchEpisodes";

import Episodes from "./Episodes";

jest.mock("../../utils/fetchEpisodes", () => jest.fn());

describe("Episodes", () => {
  beforeAll(() => {
    (fetchEpisodes as jest.Mock).mockResolvedValue(mockFetchEpisodes);
  });
  it("renders Episodes and matches snapshot", async () => {
    const { container } = render(<Episodes {...mockEpisodesProps} />);
    expect(container).toMatchSnapshot();
  });

  it("should render three EpisodesList with shared episodes data", async () => {
    render(<Episodes {...mockEpisodesProps} />);
    expect(await screen.findByText("Character 1 episodes"));
    expect(await screen.findByText("Shared episodes"));
    expect(await screen.findByText("Character 2 episodes"));
  });
});
