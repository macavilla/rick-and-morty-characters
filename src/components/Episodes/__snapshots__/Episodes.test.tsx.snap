// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Episodes renders Episodes and matches snapshot 1`] = `
<div>
  <section
    class="fixed mx-4 rounded-t-4xl bottom-0 bg-zinc-700 left-0 right-0 h-auto max-h-1/4 overflow-y-scroll border-t-4 border-zinc-900 "
  >
    <div
      class="md:flex md:justify-between md:gap-4 p-2"
    >
      <div
        class="w-full md:w-1/3"
      >
        <div
          class="px-4"
        >
          <h4
            class="text-2xl text-center"
          >
            Character 1 episodes
             
          </h4>
        </div>
      </div>
      <div
        class="w-full md:w-1/3"
      >
        <div
          class="px-4"
        >
          <h4
            class="text-2xl text-center"
          >
            Shared episodes
             
          </h4>
        </div>
      </div>
      <div
        class="w-full md:w-1/3"
      >
        <div
          class="px-4"
        >
          <h4
            class="text-2xl text-center"
          >
            Character 2 episodes
             
          </h4>
        </div>
      </div>
    </div>
  </section>
</div>
`;
